packet/src/pacer.rs

```diff
@@ -0,0 +1,131 @@
+use std::time::{Duration, Instant};
+
+pub struct LeakyBucketPacer {
+    /// Target bitrate in bits per second.
+    target_bitrate: u64,
+    /// The last time we refreshed the budget.
+    last_budget_increase: Option<Instant>,
+    /// The current budget in bytes.
+    budget: usize,
+    /// The current pacing i.e. how frequently we refresh the budget.
+    pacing: Duration,
+}
+
+pub enum Decision {
+    Emit,
+    Hold,
+}
+
+impl LeakyBucketPacer {
+    pub fn set_target_bitrate(&mut self, new_rate: u64) {
+        self.target_bitrate = new_rate;
+    }
+
+    pub fn next_interval(&self) -> Option<Instant> {
+        self.last_budget_increase.map(|l| l + self.pacing)
+    }
+
+    pub fn decide(&mut self, data: &[u8], now: Instant) -> Decision {
+        self.maybe_update_budget(now);
+        dbg!(data.len(), self.budget);
+
+        if data.len() <= self.budget {
+            self.budget -= data.len();
+            Decision::Emit
+        } else {
+            Decision::Hold
+        }
+    }
+
+    fn new(target_bitrate: u64, pacing: Duration) -> Self {
+        Self {
+            target_bitrate,
+            last_budget_increase: None,
+            budget: 0,
+            pacing,
+        }
+    }
+
+    fn maybe_update_budget(&mut self, now: Instant) {
+        let elapsed = self
+            .last_budget_increase
+            .map(|l| now - l)
+            .unwrap_or(self.pacing);
+
+        if elapsed >= self.pacing {
+            let budget = Self::calculate_budget(self.target_bitrate, elapsed);
+            self.budget = budget;
+            self.last_budget_increase = Some(now);
+            debug!(
+                "Budget increase, new budget: {}, elapsed: {:?}",
+                self.budget, elapsed
+            );
+        }
+    }
```

Comment:
I think we should keep track of whether any of the budget actually was used at all. So that we don't produce a timeout if we don't need to.

Reply:
So what I'm thinking is, imagine if we have only audio. It's quite likely we get "gaps", like below. I.e. between two pace, there is no data flowing.

I was thinking if we reach  a pace line and _NO_ budget was used during the previous interval, there is no point in scheduling any new pace, because we can just wait for the next packet and start pacing again then.

```
----- pace 5ms
packet1
packet2
packet3
.
---- pace 5ms
.
.
.
.
---- pace 5ms
.
.
.
packet 4
```

Reply:
Okay, I'll think about it and see if I can figure out a way to achieve it. Key thing is we don't want to use the budget that was unused in a previous interval when this happens I think.

An early versions of this code had this property that the pacer could essentially "save up" budget, which is not accurate. You only get `X` bytes in a given time span, if you don't use it it's gone

Reply:
Yeah. Not saving budget. Just stop pacing if previous budget was totally unused.

Reply:
libWebRTC uses the concept of debt instead of a budget which seems much nicer as a way to model this. You incur debt when sending and having excessive outstanding debt influences when you are allowed to send again.